infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironmentFactory.java

```diff
@@ -121,6 +129,32 @@ protected KubernetesEnvironment doCreate(
       }
     }
 
+    if (deployments.size() + pods.size() > 1) {
```

Comment:
This part significantly differs from the rest of the code of this method. Does it make sense to move it into a separate method and call that method here?

Reply:
It makes sense, but this block changes several collections and code looks not too elegant with putting them as arguments and expect that method will modify them.
Do you think it looks good enough or maybe you have an idea how to rewrite it?
```java
  ...
  
  if (deployments.size() + pods.size() > 1) {
      mergePods(pods, deployments, services);
  }
  ...
  }
  
    private void mergePods(Map<String, Pod> pods, Map<String, Deployment> deployments, Map<String, Service> services)
      throws ValidationException {
    List<PodData> podsData =
        Stream.concat(
            pods.values().stream().map(PodData::new),
            deployments.values().stream().map(PodData::new))
            .collect(Collectors.toList());

    Deployment deployment = podMerger.merge(podsData);
    String deploymentName = deployment.getMetadata().getName();

    // provision merged deployment instead of recipe pods/deployments
    pods.clear();
    deployments.clear();
    deployments.put(deploymentName, deployment);

    // multiple pods/deployments are merged to one deployment
    // to avoid issues because of overriding labels
    // provision const label and selector to match all services to merged Deployment
    deployment
        .getSpec()
        .getTemplate()
        .getMetadata()
        .getLabels()
        .put(DEPLOYMENT_NAME_LABEL, deploymentName);
    services.values().forEach(s -> setSelector(s, DEPLOYMENT_NAME_LABEL, deploymentName));
  }
```

Reply:
This is definitely not critical and design view here is very opinionated, so feel free to leave it as is or rework, I can't request that by any means. 
Provided snipped looks better from my POV.

Reply:
Fixed. Thanks for catching it.